frontend/src/main/scala/bloop/bsp/BloopBspServices.scala

```diff
@@ -525,37 +526,65 @@ final class BloopBspServices(
   }
 
   def run(params: bsp.RunParams): BspEndpointResponse[bsp.RunResult] = {
+    def getMainClass(project: Project, state: State): Either[Exception, bsp.ScalaMainClass] = {
+      params.dataKind match {
+        // TODO replace with RunParamsDataKind.ScalaMainClass once bsp updates
+        case Some("scala-main-class") =>
+          params.data match {
+            case None =>
+              Left(new IllegalStateException(s"Missing data for $params"))
+            case Some(json) =>
+              json.as[bsp.ScalaMainClass]
+          }
+        case Some(kind) =>
+          Left(new IllegalArgumentException(s"Unsupported data kind: $kind"))
+        case None =>
+          val cmd = Commands.Run(List(project.name))
+          Interpreter.getMainClass(state, project, cmd.main) match {
+            case Right(name) =>
+              Right(new bsp.ScalaMainClass(name, cmd.args, Nil))
+            case Left(_) =>
+              Left(new IllegalStateException(s"Main class for project $project not found"))
+          }
+      }
+    }
+
     def run(
         id: BuildTargetIdentifier,
         project: Project,
         state: State
     ): Task[State] = {
       import bloop.engine.tasks.LinkTask.{linkMainWithJs, linkMainWithNative}
       val cwd = state.commonOptions.workingPath
-      val cmd = Commands.Run(List(project.name))
-      Interpreter.getMainClass(state, project, cmd.main) match {
-        case Left(state) =>
-          Task.now(sys.error(s"Failed to run main class in ${project.name}"))
+
+      getMainClass(project, state) match {
+        case Left(error) =>
+          Task.now(sys.error(s"Failed to run main class in $project due to: ${error.getMessage}"))
         case Right(mainClass) =>
           project.platform match {
             case Platform.Jvm(javaEnv, _, _) =>
-              Tasks.runJVM(state, project, javaEnv, cwd, mainClass, cmd.args.toArray, false)
+              val env = JavaEnv(javaEnv.javaHome, javaEnv.javaOptions ++ mainClass.jvmOptions)
+              val args = mainClass.arguments.toArray
+              Tasks.runJVM(state, project, env, cwd, mainClass.`class`, args, false)
             case platform @ Platform.Native(config, _, _) =>
+              val cmd = Commands.Run(List(project.name))
               val target = ScalaNativeToolchain.linkTargetFrom(project, config)
-              linkMainWithNative(cmd, project, state, mainClass, target, platform).flatMap {
-                state =>
+              linkMainWithNative(cmd, project, state, mainClass.`class`, target, platform)
```

Comment:
Why change the change:  "mainClass.`class`" ?

Reply:
previously `mainClass` was a string, now it is a `ScalaMainClass`